### DISCUSSION TITLE
Custom makeInputSelect filter

### DISCUSSION THREAD

**Discussion by kolydart** [Power-Components/livewire-powergrid#602] [Show and tell]
Hi. I've been looking for a way to create a custom makeInputSelect filter, from an array. Could not find it in documentation or in this discussion. I found a way by creating a collection. I quote it here, in case it is usefull for others.
The model is simple; `Item` model has a `state` attribute containg a values from `[-1,0,1]`. I want to show / convert state attribute in my table to text values `['Inactive', 'Pending', 'Active']` and filter it by these values. So, the powergrid file contains the following:

```
public function datasource(): ?Builder
{
  return Item::query();
}
// [...]
public function addColumns(): PowerGridEloquent
{
  return PowerGrid::eloquent()
    ->addColumn('state')
    ->addColumn('state_name', function ($item){
        $array = [
                '1'  => 'Active',
                '0'  => 'Pending',
                '-1' => 'Inactive',
            ];                
        return $array[$item->state];
    })
// [...]
Column::add()
        ->title('state')
        ->field('state_name', 'state')
        ->makeInputSelect(collect([
            ['id'=>1, 'state'=>'Active'],
            ['id'=>0, 'state'=>'Pending'],
            ['id'=>-1, 'state'=>'Inactive'],
        ]),'state','state')
        ,
```
Result: 
<img width="273" alt="Screenshot 2022-09-02 at 9 44 50 AM" src="https://user-images.githubusercontent.com/8072845/188075446-715fdb55-5966-4eea-b298-b0eaa5c6d3ff.png">

**Comment by dansysanalyst** [comment]
Hi @kolydart,

` 💪 ` Thanks for sharing. You are absolutely right, this is obscure in the documentation. 

I will push a PR to the doc and another to [Powergrid Demo](https://github.com/Power-Components/powergrid-demo) featuring a clear example.

---

My first idea was to suggest you to consider using an Enum for the states and combine it with the [makeInputEnumSelect](https://livewire-powergrid.com/#/table/column-filters?id=makeinputenumselectarray-enumcases-string-datafield-null-array-settings-) filter.

However, if this is not possible, I would take the following approach:

📌 Add a method on `User` model returning a collection with each status code and label.

```php
// File: app/Models/User.php
<?php 

use Illuminate\Support\Collection;

class User extends Authenticatable
{
    //...
    public static function statuses(): Collection
    {
        return collect(
            [
                ['status' => 1,  'label' => 'Active'],
                ['status' => 0,  'label' => 'Pending'],
                ['status' => -1, 'label' => 'Inactive'],
            ]
        );
    }
```

📌 Add a column `status_label`, which gets the first status in the collection matching the database value.

```php 
    public function addColumns(): PowerGridEloquent
    {
        return PowerGrid::eloquent()
            //...
            ->addColumn('status')
            ->addColumn('status_label', fn ($user) => User::statuses()->firstWhere('status', $user->status)['label'])
    }
```

📌 Include the column `status_label` and specify the `$displayField` to read the `label` key.

```php 
    public function columns(): array
    {
        return [
             //...
            Column::add()
                ->title('status')
                ->field('status_label', 'status')
                ->makeInputSelect(User::statuses(), 'label', 'status'),
       ];
    }
```

Resulting in:

<img width="798" alt="CleanShot 2022-09-02 at 11 12 06@2x" src="https://user-images.githubusercontent.com/79267265/188106506-cd8ff433-662a-4f21-a7d6-e5c2999139d9.png">


In this way, if you need to change, add or remove any of the statuses, you only have to modify the `User::statuses()` method.

**Reply by kolydart** [reply]
Hi, @dansysanalyst,
thanks for the reply.
You are right, your suggestion is very elegant. I sent the message as soon as I got it working with minimal bare bones structure. I guess I would turn to your approach eventually. Thanks for helping on it.
